### DISCUSSION TITLE
Comparison with Depth-Anything-V2 demo

### DISCUSSION THREAD

**Discussion by gituser123456789000** [nagadomi/nunif#149] [General]
I pushed for Tiling.. and I assume we got some sort of BoostingM0nocularDepth added in.. but it's causing issues. We're getting worse maps than default Depth-Anything v2

Comparing side by side, back and forth, the huggingface demo results vs iw3's depth map output at the same settings.. there are many issues and all bad

I used a 1000x1333 image. The huggingface demo gave me this native resolution as the result... but putting in 1000 resolution in iw3 gave a result of 1008x1330.. which is 8 pixels wider, but 3 pixels SHORTER (instead of slightly taller to compensate for the extra width that was introduced. That right there is a big problem. Now the depth map doesn't align with the original image.

Beyond that.. the maps look very similar.. you can see they're made from the same model.. but the shading, the colors, the graident is off in many areas. The nose is slightly brighter, which means it will stick out more.. the eyes have more definition, which looks better at first glance, but that can mean the eyes will be sunken in.. other areas, where the body should be relatively uniform or just slightly different in shade, it's instead too much difference in gradient, so basic body or object shape is distorted.

As far as BoostingMonocularDepth (assuming we got this or some similar adaptation of it).. I never got an impressive looking result from that when I downloaded it locally and used the included MiDaS v2 and Leres conversions. Their demo looks impressive, but I got nothing impressive from it. I chalked it up to MiDaS 2 being an old model, but I think BoostingMonocularDepth may just not be great or accurate. It may produce nice looking detail in maps, but not accuracy.

I pushed for Tiling, but in this case, whatever implementation this is.. it's not working as I hoped.

We need default, native resolution, straight from the Depth-Anything v2 models. It looks impressive, because it seems to already be the best.. but iw3 is currently altering the output to be something that's less accurate.

Native resolution, direct result should be the default option.. or some check box to process in native resolution, direct from the model... and whatever scaling / tiling / BoostingMonocularDepth type action is happening, should be an option at best, but really, we should consider removing it if it's only causing inaccuracies.

I'm just delivering factual feedback to be clear. We appreciate your work very much, nagadomi. Thank you for attempting to implement something here as an improvement, but in this case, it's not performing as we may have expected it to.

**Comment by nagadomi** [comment]
First, iw3 does not use anything like BoostingMonocularDepth.
There may be some differences in implementation, but the process is basically the same as in the demo.

Regarding depth resolution,
DepthAnything model requires that the input size is a multiple of 14 for height and width.
1000 and 1333 are not multiples of 14, so it is resized to 1008x1330 before input to DepthAnything model.
When the depth is actually used, it is resized back to its original resolution 1000x1333.
The demo simply outputs the depthmap after this resize back, so it matches the original resolution.
iw3 does not resize back to the original resolution until it is actually used. Because it is a waste of space. In export feature, it wastes several to dozens of times the storage size.

Regarding the different gradients of depth,
the demo outputs 8-bit grayscale (256 level) and iw3 outputs 16 grayscale (65536 level).
At the moment I think that's the difference, but I'm sure many changes have been made recently so I will look into it again in more detail.

EDIT:
If you are testing on the official demo, it is always running with Depth Resolution=518 (and resize back to the original resolution). So if you want to compare, you need to use Depth Resolution=518 in iw3.

**Reply by gituser123456789000** [reply]
> First, iw3 does not use anything like BoostingMonocularDepth. There may be some differences in implementation, but the process is basically the same as in the demo.
> 
> Regarding depth resolution, DepthAnything model requires that the input size is a multiple of 14 for height and width. 1000 and 1333 are not multiples of 14, so it is resized to 1008x1330 before input to DepthAnything model. When the depth is actually used, it is resized back to its original resolution 1000x1333. The demo simply outputs the depthmap after this resize back, so it matches the original resolution. iw3 does not resize back to the original resolution until it is actually used. Because it is a waste of space. In export feature, it wastes several to dozens of times the storage size.
> 
> Regarding the different gradients of depth, the demo outputs 8-bit grayscale (256 level) and iw3 outputs 16 grayscale (65536 level). At the moment I think that's the difference, but I'm sure many changes have been made recently so I will look into it again in more detail.
> 
> EDIT: If you are testing on the official demo, it is always running with Depth Resolution=518. So if you want to compare, you need to use Depth Resolution=518 in iw3.

518 did produce a very much more similar result as the official demo, but not 100% exactly same, there's a slight shading variation..

the output is 518x686, whereas the huggingface demo outputs the original 1000x1333 size depth map. iw3's map is still wider in this case, which can be seen when switching back and forth..

the aspect ratios are
1000x1333 = 1.333 : 1
518 x 686  = 1.324 : 1

so the depth map seems to still not align to the original image properly

**Reply by gituser123456789000** [reply]
I'm thinking this different aspect ratio output may or may not be contributing to edge issues in some images/videos.. likely seen more at higher strength and congergence

**Reply by gituser123456789000** [reply]
An unrelated idea for possible improvement, to eliminate color banding in the depth maps, would be to output the depth maps in 32bit .tiff as the fork for TilingZoeDepth GUI added support for. I'm not sure if this is possible in iw3 or with Depth-Anything.. I don't know how all of that works.., but that's an idea to test.

**Reply by gituser123456789000** [reply]
iw3 output is not exact as the huggingface demo..
I tested this random selena gomez image (just because it was a square image with some complexity...)

The original image is 3840x3840.. I also made a small version to match the 518 training/processing number.. just to see what the results would be. The iw3 shading does not match. This looks to be causing things I mentioned in a discord yesterday. Sunken in necks, sunken in eyes, etc. iw3 depth maps look to be slightly more detailed, and that looks good to our eyes, but it's bad for results.

I used 518x518 versions to rule out differences in resizing/aspect ratios

Here are the results of the following test inputs:
518x518 iw3
518x518 huggingface demo
3840x3840 iw3 (it output as 518x518)
3840x3840 huggingface demo (it output as 3840x3840)

![518 iw3](https://github.com/nagadomi/nunif/assets/156575428/ebd3f796-63c9-4ab9-8982-6b5e344a77c6)
![518 huggingface demo](https://github.com/nagadomi/nunif/assets/156575428/1a106f98-3ae2-4e9b-a0f0-4a82ba7699b5)
![3840 iw3](https://github.com/nagadomi/nunif/assets/156575428/462522ed-9f34-4552-8075-3e51ca8924de)
![3840 huggingface demo](https://github.com/nagadomi/nunif/assets/156575428/5055266a-c710-4d3b-8d2d-4f31efe5b075)

Comparing you'll likely have to save to your PC and insta-switch back and forth.. you'll notice differences in shading, and some details. iw3 isn't 1:1 the same output as Depth-Anything directly.. even when using exactly 518x518 inputs

**Reply by nagadomi** [reply]
>shading

see this comment `>the demo outputs 8-bit grayscale (256 level) and iw3 outputs 16 grayscale (65536 level).`

>aspect ratios

It is impossible to maintain a perfect aspect ratio because width and height must be multiples of 14 and the original resolution is not.

>3840x3840

The difference in this case is probably due to differences in downscaling algorithms.
The demo uses OpenCV's bicubic downscaling and iw3 uses PyTorch's bilinear downscaling. So the input images after resizing are slightly different.

**Reply by gituser123456789000** [reply]
I did more testing with a different image and used 3DCombined with the iw3 depth map results (upscaled to the original image's resolution and aspect ratio in paint.net (this step is necessary to not get an error in 3DCombine.. because the original image and depth map resolution must match)) vs the huggingface demo depth map results

this way a third party app is doing the putting together part for both...

The results are not absolutely 100% identical, but they essentially are. There are very slight differences noticable when insta-switching back and forth, but I'd call it 99% the same. Some section of hair may be a cm closer.. an eye may be a millimeter difference in depth, etc. The differences were very minimal. I would say I probably preferred the direct huggingface demo depth map results (just judging by this 1 test image).. but again, it was 99 vs 100.. basically the same

Everyone converting, I highly suggest sticking with 518 as your Depth Resolution to get the most true-to-original results.. with Depth-Anything v2 models at least

**Reply by gituser123456789000** [reply]
> > shading
> 
> see this comment `>the demo outputs 8-bit grayscale (256 level) and iw3 outputs 16 grayscale (65536 level).`
> 
> > aspect ratios
> 
> It is impossible to maintain a perfect aspect ratio because width and height must be multiples of 14 and the original resolution is not.
> 
> > 3840x3840
> 
> The difference in this case is probably due to differences in downscaling algorithms. The demo uses OpenCV's bicubic downscaling and iw3 uses PyTorch's bilinear downscaling. So the input images after resizing are slightly different.

Does the iw3 depth map result resolution re-scale to the original resolution like the huggingface demo does before iw3 processes it?

for example the aspect ratios I mentioned..

1000x1333 = 1.333 : 1 (original image)
518 x 686 = 1.324 : 1 (iw3 depth map output)

does iw3 first upscale and resize / re-aspect ratio the depth map image to the original image's resolution before processing... 

because if it just upscales, for example, it would be combining a 1000x1333 image with a 1000x1324 depth map.. so results would be slightly off.. again, just thinking this may effect the accuracy of the output and edges.... in this case, just half of 1%, but that's still something..

**Reply by nagadomi** [reply]
>Does the iw3 depth map result resolution re-scale to the original resolution like the huggingface demo does before iw3 processes it?>

The details of this part is complicated for implementation reasons, but basically it works as follows.
```python
original_image = input_frame# 1000x1333
# infer depth
model_input_image = resize_for_model_input(original_image) # 518x686
depth = infer_depth(model_input_image) # output depth size = 518x686
# generate sbs
depth = resize_back_for_original(depth) # depth size = 1000x1333
sbs = generate_sbs(original_image, depth) # original_image and depth are the same size
```

This has a pixel alignment problem as you mentioned.
But when depth is much smaller than the original resolution, the upscaling factor for depth will be larger, and low-quality upscaling will be a much bigger problem I think.
So at this point, I am not concerned about this problem.

Not long ago, the quality of depth was not good enough to make me nervous about this problem :|

**Reply by gituser123456789000** [reply]
> > Does the iw3 depth map result resolution re-scale to the original resolution like the huggingface demo does before iw3 processes it?>
> 
> The details of this part is complicated for implementation reasons, but basically it works as follows.
> 
> ```python
> original_image = input_frame# 1000x1333
> # infer depth
> model_input_image = resize_for_model_input(original_image) # 518x686
> depth = infer_depth(model_input_image) # output depth size = 518x686
> # generate sbs
> depth = resize_back_for_original(depth) # depth size = 1000x1333
> sbs = generate_sbs(original_image, depth) # original_image and depth are the same size
> ```
> 
> This has a pixel alignment problem as you mentioned. But when depth is much smaller than the original resolution, the upscaling factor for depth will be larger, and low-quality upscaling will be a much bigger problem I think. So at this point, I am not concerned about this problem.
> 
> Not long ago, the quality of depth was not good enough to make me nervous about this problem :|

Could you try to implement a 32bit processing option like TilingZoeDepth GUI fork added to TilingZoeDepth project?

https://github.com/n8ventures/TilingZoeDepth_GUI/releases/tag/3.1.0

**Reply by nagadomi** [reply]
I think the difference between the 16-bit and 32-bit modes of the TilingZoeDepth GUI is not due to differences in output formats.
The "Result" image on the release page is **16-bit PNG**. This means that the difference is not between 32-bit and 16-bit output formats.
Also, iw3 uses 32-bit float (not int) for processing.

**Reply by gituser123456789000** [reply]
> I think the difference between the 16-bit and 32-bit modes of the TilingZoeDepth GUI is not due to differences in output formats. The "Result" image on the release page is **16-bit PNG**. This means that the difference is not between 32-bit and 16-bit output formats. Also, iw3 uses 32-bit float (not int) for processing.

The output result from the program is 32bit .tiff with the opposite grayscale

**Reply by gituser123456789000** [reply]
I'd post an example file, but we can't upload .tiff here..

here's a .png conversion of it though, that looks the same:
![dino_TiledZoeDepth 32bit](https://github.com/nagadomi/nunif/assets/156575428/d18f2366-f650-4700-8b55-41e31db2592a)

Here are further examples of iw3 default (ZoeDepth N default), iw3 ZoeN 512 resolution (I tried 768 but it's very messed up looking), TiledZoeN 16bit, and TiledZoeN 32bit (inverted colors to match the white front / black rear of others)

![dino default upscaled to original res](https://github.com/nagadomi/nunif/assets/156575428/5b5bd124-1b80-4a68-82f6-147ad01ce303)
![dino 512 upscaled to original res](https://github.com/nagadomi/nunif/assets/156575428/53aadd95-ade9-472b-8a95-da5cc2cdd2c5)
![dino_TiledZoeDepth 16bit](https://github.com/nagadomi/nunif/assets/156575428/b7f713d7-43dd-4792-a84c-e28f12f16529)
![dino_TiledZoeDepth 32bit inverted](https://github.com/nagadomi/nunif/assets/156575428/c8664d3d-48f6-445f-a223-cbcda5e7f109)

I don't love the accuracy of ZoeDepth to begin with, so I'd ignore much of the inaccuracies in any of the results, but that's all I have available for tiling. I'd rather be seeing the results of tiling Depth-Anything models.

I don't know how much, if any, more Tiling is done with the 16bit and 32bit conversions in that program, but you can see more detail in the gradient in the 32bit result

**Reply by gituser123456789000** [reply]
As far as just the banding I mentioned in 16bit.. here's 32bit vs 16bit example image showing the banding difference.

![32bitvs16bit](https://github.com/nagadomi/nunif/assets/156575428/fb854121-0a0f-474b-9e97-0e69b401ee0b)

I'm sure you are likely very aware already, but I have noticed the banding in iw3 maps at 16bit, so I think 32bit output and processing could be an area of improvement. I saw you mentioned there's a 32bit processing, but I don't understand exactly how that works.. I can just see the 16bit map...

So are you saying everything is processed at 32bit.. but the depth maps are output at 16bit and other results like I see anaglyphs are 24bit? I'm going to have to run tests and see if I see banding in final results and not just in the depth maps

**Reply by nagadomi** [reply]
>here's 32bit vs 16bit example 

It is not 16bit single channel. It may be "16bit color". If so, the red channel is only 5bit.

16bit grayscale is 65536 levels.
Theoretically, a gradient image of width=65536px can be created without banding.
However, it cannot be displayed correctly on PC with 24/32-bit color settings.

If you see banding in a 16bit depth image, it is not due to 16bit grayscale, it is due to other causes, such as block artifact problem in the model.

**Reply by nagadomi** [reply]
I realized it today, but the difference between 8-bit grayscale, 16-bit grayscale, and 32-bit grayscale may not be visible on the PC monitor.

**Reply by gituser123456789000** [reply]
ok. I'm running comparisons of iw3 vs their sample images now and things seem to line up with the banding and pixelated edges in their samples compared to what comes out of iw3

![22984102999_aa45c5cf50_k_v2](https://github.com/nagadomi/nunif/assets/156575428/a11411b8-c34f-4f9d-a0fa-97d2a67bcfbf)

banding in the cow, the vehicle/cart in the middle. pixelated edges. Theirs obviously are colormapped, but it lines up to iw3's output. I might have found some that don't match, but I'm still verifying..

**Reply by gituser123456789000** [reply]
All good.. v2 is smarter than I thought.

![40277550594_7e379065f3_k_img](https://github.com/nagadomi/nunif/assets/156575428/8ea190a5-5aef-49fe-9470-979bb9896bc4)
![40277550594_7e379065f3_k_img](https://github.com/nagadomi/nunif/assets/156575428/0d3f3aac-b0da-45cc-8968-e5a03fc97168)

I saw the depth map and first thought.. that can't be right.. but yeah, it's smart and knew that's a glass building

<img width="617" alt="Capture" src="https://github.com/nagadomi/nunif/assets/156575428/910c5c66-9480-4a99-8f56-c7108e3e80c4">

**Reply by nagadomi** [reply]
Its colormap has only 256 color palette.

**Reply by gituser123456789000** [reply]
> I realized it today, but the difference between 8-bit grayscale, 16-bit grayscale, and 32-bit grayscale may not be visible on the PC monitor.

I'm still in bed, so I haven't looked fully into this, but here they apparently eliminated the banding 

https://www.reddit.com/r/StableDiffusion/comments/1dmypej/making_3d_basreliefs_with_depth_anything_v2_16bit/

**Reply by nagadomi** [reply]
iw3 already outputs in 16bit. If you check with Blender in the same way you will see that there is no banding.

**Reply by nagadomi** [reply]
I checked with Blender. Note that this does not use Displacement, but simply uses Bump (I can see the difference enough).
Left is demo ouptut(8bit), Right is iw3 Export output (16bit). Click to view larger image.
![depth-8bit-vs-16bit](https://github.com/nagadomi/nunif/assets/287255/ec984c78-5091-46e4-ab8c-f0905a6eaf72)

Both depthmaps look the same on the PC monitor.
demo outout(8bit)
![depth-demo](https://github.com/nagadomi/nunif/assets/287255/843892db-f789-49fb-bfbf-8838b9b1ef24)

iw3 Export output (16bit)
![depth-iw3](https://github.com/nagadomi/nunif/assets/287255/740bd672-1e9a-416c-b9be-ffa1e08bcfab)

**Reply by nagadomi** [reply]
I have changed the title to the suitable title. Please edit if you have an alternate title.

**Reply by gituser123456789000** [reply]
> I have changed the title to the suitable title. Please edit if you have an alternate title.

all good.. that's a fine title

I'm just always looking for better quality if it's possible. Like right now I want to see how good that Giant version of the model is.

I know you said iw3 was already being processed at a higher bitrate, but I have to learn and understand why I'm seeing what I'm seeing.

This way in Blender you posted, shows it nice and clearly.

Our monitors are lying to us by only being 8bit in most cases.. so it's hard to initially believe our eyes are wrong when just looking at the depth maps, but there's extra, additional hidden information in the higher bit files that we just can't see in that way, unless we do certain projects, like shown in Blender, or until we maybe get higher bit depth monitors. I think 10bit monitors or VR headsets are probably already a thing... or will be soon. It's also interesting to know that content we make now might already have the ability to look better in the future, on more advanced displays.
